### PR TITLE
Hardware interface for adcs

### PR DESCRIPTION
**Pull request by QuocTrungTran**
Implemented the hardware interface for adcs. Most of the functions are trivial, but the most important thing is the get housekeeping data function.
I also added a function to get satellite position (LLH) since we need it for housekeeping data. 
On other note, I stick with the type convention of each housekeeping data from the previously developed code, although the type according to the housekeeping data sheet is different. For example, the Estimated Angular Rate should be in int16_t, but I used xyz struct instead.

### CODE REVIEW

File: equipment_handler/inc/drivers/sci.h [hunk 1]
```diff
@@ -1,168 +1,168 @@
-/** @file sci.h
```
Comment (QuocTrungTran) [inline equipment_handler/inc/drivers/sci.h:1]: Not sure how this is a change. I didn't touch this file and they are identical anyways.

File: hardware_interface/include/adcs.h [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ADCS_returnState HAL_ADCS_set_ywheel_config(ywheel_ctrl_config params);
+ADCS_returnState HAL_ADCS_set_rwheel_config(rwheel_ctrl_config params);
+ADCS_returnState HAL_ADCS_set_tracking_config(track_ctrl_config params);
+ADCS_returnState HAL_ADCS_set_MoI_mat(moment_inertia_config cell);
+ADCS_returnState HAL_ADCS_set_estimation_config(estimation_config config);
+ADCS_returnState HAL_ADCS_set_usercoded_setting(usercoded_setting setting);
+ADCS_returnState HAL_ADCS_set_asgp4_setting(aspg4_setting setting);
+ADCS_returnState HAL_ADCS_get_full_config(adcs_config* config);
+
+ADCS_returnState HAL_ADCS_getHK(ADCS_HouseKeeping* adcs_hk);
```
Comment (wagnerd97) [inline hardware_interface/include/adcs.h:192]: Has this been implemented or is just the prototype made?

File: equipment_handler/inc/drivers/sci.h (hunk 1 above)
Comment (ArashYzd) [reply equipment_handler/inc/drivers/sci.h:1]: It's because of moving it.

**Review by ArashYzd — COMMENTED** [review]
1. Did you find a way to test the functions?
2. How are you going to test the service if HAL functions just call handler ones?
3. Are you going to write a subservice for each handler function? Why would we need a HAL function if that's not gonna be used for the service?

File: hardware_interface/include/adcs.h (hunk 2 above)
Comment (ArashYzd) [reply hardware_interface/include/adcs.h:192]: See the .c file.

Comment (QuocTrungTran) [reply hardware_interface/include/adcs.h:192]: yes, the actual implementation is in the adcs.c file

**Comment by QuocTrungTran** [comment]
> 1. Did you find a way to test the functions?
> 2. How are you going to test the service if HAL functions just call handler ones?
> 3. Are you going to write a subservice for each handler function? Why would we need a HAL function if that's not gonna be used for the service?

1. Because I am just calling the handler functions and they were fully tested before so I think it would be fine. But I can dig deeper to test them as well as the new housekeeping function. I will find a way to resolve the ceedling command not running. 
2. I'm also confused at this. Looking at other hardware interfaces, they also call the handler ones. For example, the sband_software. At first, I dont understand why we did it because it is just a repetitive and unnecessary call to handler functions but even you confirmed with me the hardware intefaces call handler ones. Maybe I understand it wrong? 
3. see above

**Review by wagnerd97 — APPROVED** [review]

**Comment by ArashYzd** [comment]
> > 1. Did you find a way to test the functions?
> > 2. How are you going to test the service if HAL functions just call handler ones?
> > 3. Are you going to write a subservice for each handler function? Why would we need a HAL function if that's not gonna be used for the service?
> 
> 1. Because I am just calling the handler functions and they were fully tested before so I think it would be fine. But I can dig deeper to test them as well as the new housekeeping function. I will find a way to resolve the ceedling command not running.
> 2. I'm also confused at this. Looking at other hardware interfaces, they also call the handler ones. For example, the sband_software. At first, I dont understand why we did it because it is just a repetitive and unnecessary call to handler functions but even you confirmed with me the hardware intefaces call handler ones. Maybe I understand it wrong?
> 3. see above

There's a history behind why sband and uhf are like that. It's from the old architecture where the handler codes did not exist in the obc_software. Having a HAL function matches our software architecture and has some benefits. One is locally saving the values on obc in case the hardware is unavailable. That attribute also allowed for assigning dummy values that can be helpful for testing the service. So they are not "just" calling the handler functions. Another use is generating commands like getHK that calls different functions. My argument is though, why each handler has a HAL function if above justifications do not apply to them?

**Comment by QuocTrungTran** [comment]
> > > 1. Did you find a way to test the functions?
> > > 2. How are you going to test the service if HAL functions just call handler ones?
> > > 3. Are you going to write a subservice for each handler function? Why would we need a HAL function if that's not gonna be used for the service?
> > 
> > 
> > 
> > 1. Because I am just calling the handler functions and they were fully tested before so I think it would be fine. But I can dig deeper to test them as well as the new housekeeping function. I will find a way to resolve the ceedling command not running.
> > 2. I'm also confused at this. Looking at other hardware interfaces, they also call the handler ones. For example, the sband_software. At first, I dont understand why we did it because it is just a repetitive and unnecessary call to handler functions but even you confirmed with me the hardware intefaces call handler ones. Maybe I understand it wrong?
> > 3. see above
> 
> There's a history behind why sband and uhf are like that. It's from the old architecture where the handler codes did not exist in the obc_software. Having a HAL function matches our software architecture and has some benefits. One is locally saving the values on obc in case the hardware is unavailable. That attribute also allowed for assigning dummy values that can be helpful for testing the service. So they are not "just" calling the handler functions. Another use is generating commands like getHK that calls different functions. My argument is though, why each handler has a HAL function if above justifications do not apply to them?

I actually just followed the convention we current have and sorry I didn't know about that history. Yes I agree with you that these new HAL functions are pointless. So I guess we should get rid of them? Perhaps only keep the housekeeping function. I dont know what else I should do here. Enlighten me.

**Comment by ArashYzd** [comment]
I think at this point only the housekeeping and the ones you need for the service should be enough. Though, keep that commit somewhere locally in case you need to borrow more.
You might need to come back and add more stuff to the functions you'll keep for services. So, until then, unit tests are not necessary. You can test the full path later.
For HK, too, ask Dustin how to verify it. Maybe his review on this PR is enough. I'm not sure.

File: hardware_interface/include/adcs.h [hunk 3]
```diff
@@ -30,6 +30,143 @@ typedef struct __attribute__((packed)) {
...
+typedef struct __attribute__((packed)) {
+    uint32_t time;
+    uint8_t event_id;
+    uint8_t event_param;
+} ADCS_last_logged_event;
+
+typedef struct __attribute__((packed)) {
+    bool format_busy;
+    bool erase_all_busy;
+} ADCS_SD_format_progress;
```
Comment (ArashYzd) [inline hardware_interface/include/adcs.h:63]: I'm not sure how packed structs with bools would work but probably better not to copy this struct to the packet. I think parsing 0x0101 is better than 0x03.

File: hardware_interface/include/adcs.h [hunk 4]
```diff
@@ -30,6 +30,143 @@ typedef struct __attribute__((packed)) {
...
+} ADCS_unix_t;
+
+typedef struct __attribute__((packed)) {
+    uint16_t uptime;
+    uint8_t flags_arr;
+} ADCS_bootloader_state;
+
+typedef struct __attribute__((packed)) {
+    uint8_t index;
+    bool busy;
```
Comment (ArashYzd) [inline hardware_interface/include/adcs.h:126]: same here. Doesn't it shift the ones after by one bit?

File: hardware_interface/include/adcs.h (hunk 3 above)
Comment (QuocTrungTran) [reply hardware_interface/include/adcs.h:63]: Sure I can separate them

**Review by ArashYzd — APPROVED** [review]

**Pull request merged**